### PR TITLE
Lathe Refactor and ECS

### PR DESCRIPTION
**Pull request by EmoGarbage404**
### Material Storage
Material storage component and materials are now in shared. Material storage has access on it so you must use the plethora of kind helper functions i have supplied. Other than that, it remains mostly unchanged, besides being ECS'd.

### Lathe Database
LatheDatabase, ProtolatheDatabase, and all the related database components have been removed. In their place, lathe component has gained two lists: StaticRecipes and DynamicRecipes.

Static recipes replace autolathe recipes: they're always available to print. DynamicRecipes are like Protolathe recipes. They serve as a guide for what technology the lathe can unlock. Code note here: the DynamicRecipe list is not authoritative, so checking if it's on that list is necessary.

The list of available recipes, as such, is no longer static. Subscribing to the LatheGetRecipesEvent allows a system to add additional recipes to a lathe.

### Lathe Sever-side
Basically just a lot of cleanup here. Besides the fact that materials are now used on queue rather than print, it's basically the same. Otherwise it's just ECS'd

### Lathe Client-side
The ui has been converted to xaml. The itemlist that recipes used is now just a scroll container with custom controls.

**Changelog**

:cl:
- fix: Fixed various weird issues with lathe behavior that would result in them getting stuck.

### CODE REVIEW

File: Content.Server/Lathe/Components/LatheProducingComponent.cs [hunk 1]
```diff
@@ -1,24 +1,23 @@
...
-        /// Remaining production time, in seconds.
-        /// </summary>
-        [DataField("timeRemaining", required: true)]
-        public float TimeRemaining;
-    }
+    /// <summary>
+    /// Remaining production time, in seconds.
+    /// </summary>
+    [ViewVariables(VVAccess.ReadWrite)]
+    public float TimeRemaining;
```
Comment (metalgearsloth) [inline Content.Server/Lathe/Components/LatheProducingComponent.cs:21]: After much back and forth you should just use the finish time. It won't handle pausing properly until we get some niceties around it (i.e. timespanserializer + a way to check how long an entity has been paused) but it will save more work later.

File: Content.Shared/Lathe/LatheComponent.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    public sealed class LatheComponent : Component
+    {
+        /// <summary>
+        /// All of the recipes that the lathe has by default
+        /// </summary>
+        [DataField("staticRecipes", customTypeSerializer: typeof(PrototypeIdListSerializer<LatheRecipePrototype>))]
+        public readonly List<string> StaticRecipes = new();
+
+        /// <summary>
+        /// All of the recipes that the lathe is capaable of researching
```
Comment (metalgearsloth) [inline Content.Shared/Lathe/LatheComponent.cs:18]:
```suggestion
        /// All of the recipes that the lathe is capable of researching
```

File: Content.Server/Lathe/LatheSystem.cs [hunk 3]
```diff
@@ -54,217 +57,172 @@ public override void Update(float frameTime)
...
-
-            // TODO: You can probably split this part off of lathe component too
-            if (!TryComp<LatheComponent>(uid, out var lathe))
-                return;
-
-            // We need the prototype to get the color
-            _prototypeManager.TryIndex(lastMat, out MaterialPrototype? matProto);
+                Recipes = component.StaticRecipes
+            };
+            RaiseLocalEvent(component.Owner, ev, true);
```
Comment (metalgearsloth) [inline Content.Server/Lathe/LatheSystem.cs:116]: Does it need to be raised as broadcast?

File: Content.Server/Lathe/LatheSystem.cs [hunk 4]
```diff
@@ -54,217 +57,172 @@ public override void Update(float frameTime)
...
+            if (!Resolve(uid, ref component))
                 return false;
-
-            if (!this.IsPowered(uid, EntityManager))
+            if (HasComp<LatheProducingComponent>(uid) || !component.Queue.Any())
                 return false;
 
-            var recipeId = component.Queue[0];
+            var recipe = component.Queue.First();
+            component.Queue.RemoveAt(0);
```
Comment (metalgearsloth) [inline Content.Server/Lathe/LatheSystem.cs:152]: Probably better to just make it a stack as that's what it functions as IMO.

File: Content.Shared/Materials/SharedMaterialStorageSystem.cs [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+    /// are allowed to be inserted into the materialStorage.
+    /// </summary>
+    /// <param name="uid"></param>
+    /// <param name="component"></param>
+    public void UpdateMaterialWhitelist(EntityUid uid, MaterialStorageComponent? component = null)
+    {
+        if (!Resolve(uid, ref component, false))
+            return;
+        var ev = new GetMaterialWhitelistEvent(uid);
+        RaiseLocalEvent(uid, ev, true);
```
Comment (metalgearsloth) [inline Content.Shared/Materials/SharedMaterialStorageSystem.cs:200]: As above.

File: Content.Shared/Materials/SharedMaterialStorageSystem.cs [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+        if (!Resolve(uid, ref component))
+            return false;
+        if (!CanChangeMaterialAmount(uid, materialId, volume, component))
+            return false;
+        if (!component.Storage.ContainsKey(materialId))
+            component.Storage.Add(materialId, 0);
+        component.Storage[materialId] += volume;
+
+        if (!_defferedMaterialChangeEvent.Contains(component))
+            _defferedMaterialChangeEvent.Add(component);
```
Comment (metalgearsloth) [inline Content.Shared/Materials/SharedMaterialStorageSystem.cs:130]: Do these need to be deferred? Storing data on systems become funny with serialization.

File: Content.Shared/Materials/SharedMaterialStorageSystem.cs [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+        args.State = new MaterialStorageComponentState(component.Storage, component.MaterialWhiteList);
+    }
+
+    private void OnHandleState(EntityUid uid, MaterialStorageComponent component, ref ComponentHandleState args)
+    {
+        if (args.Current is not MaterialStorageComponentState state)
+            return;
+
+        component.Storage = state.Storage;
+        component.MaterialWhiteList = state.MaterialWhitelist;
```
Comment (metalgearsloth) [inline Content.Shared/Materials/SharedMaterialStorageSystem.cs:37]: This will store refs to the state data which means it won't be handled correctly for prediction, you need to copy the data (using linq is fine). I often get caught out by this and electro has to point it out.

File: Content.Server/Lathe/LatheSystem.cs [hunk 8]
```diff
@@ -54,217 +57,172 @@ public override void Update(float frameTime)
...
+            }
+
+            foreach (var (comp, lathe) in EntityQuery<LatheProducingComponent, LatheComponent>())
             {
                 comp.TimeRemaining -= frameTime;
 
                 if (comp.TimeRemaining <= 0)
-                    FinishProducing(comp.Owner, comp);
+                    FinishProducing(comp.Owner, lathe);
             }
```
Comment (metalgearsloth) [inline Content.Server/Lathe/LatheSystem.cs:74]: Wouldn't it be better to add LatheProducingComponent if the queue ever goes > 0 then check powered in the bottom method?

**Review by metalgearsloth — DISMISSED** [review]

File: Content.Server/Lathe/LatheSystem.cs (hunk 4 above)
Comment (EmoGarbage404) [reply Content.Server/Lathe/LatheSystem.cs:152]: As far as I know, it cannot be a stack because the recipes are taken off the front of the queue and added to the end. From what i can see, stacks only support popping off the top.

File: Content.Shared/Materials/SharedMaterialStorageSystem.cs (hunk 6 above)
Comment (EmoGarbage404) [reply Content.Shared/Materials/SharedMaterialStorageSystem.cs:130]: this was done to cut down on a few extra ui messages and updates, but yea, it probably doesn't matter and it was done like that before.

File: Content.Server/Lathe/LatheSystem.cs [hunk 9]
```diff
@@ -286,56 +241,62 @@ private void UpdateRunningAppearance(EntityUid uid, bool isRunning)
...
+            {
+                if (TryComp<LatheProducingComponent>(uid, out var prodcomp))
+                    RemComp(uid, prodcomp);
+            }
+            else if (component.Recipe != null)
+            {
+                EnsureComp<LatheProducingComponent>(uid);
+            }
+            UpdateRunningAppearance(uid, args.Powered && HasComp<LatheProducingComponent>(uid));
+        }
```
Comment (metalgearsloth) [inline Content.Server/Lathe/LatheSystem.cs:261]: RemComp returns a bool so shouldn't need TryComp

Also couldn't you just check if there's a queue and run TryStartProducing instead of in the update?

**Comment by mirrorcult** [comment]
explanation of what exactly the refactor is doing in the PR body would be nice

**Comment by EmoGarbage404** [comment]
wrote up a proper description, updated tests, and completed review. also thoroughly tested in game.
